### DISCUSSION TITLE
Thanks, and Filament Spool Change Question

### DISCUSSION THREAD

**Discussion by mtinman** [mdziekon/octoprint-spoolman#33] [Q&A]
mdziekon: Thank You for developing this plugin, and I was wondering, is there a means to change the spool values when doing color/filament changes during a print with your plugin? I had seen some mention of this here: https://github.com/Donkie/Spoolman/discussions/272, and was wondering if this might be incorporated into your plugin? I'm pretty sure this would be a great option if possible, so that spools could be better managed during prints that involve color/filament changes during the print. I hope I am being clear with my request, if not, please let me know...

**Comment by mdziekon** [comment]
Hi @mtinman thanks for reaching out.

Currently, there's no simple way to **manually** change a spool mid-print, but M600 should be somewhat supported, with just a bit of user interaction.

When dealing with filament changes embedded in GCODE (eg. M600 color changes), and as long as your printer / octoprint pauses the print once that GCODE command is reached, this situation **should be supported** - the pluggin commits usage data on print pause event, after that you can safely change selected spool, and then resume the print. You could also pause your print manually, as that's virtually the same thing.

However, manually changing just the selected spool, without pausing the print, is not supported right now, and actually should be considered a bug - the plugin will politely change the selection, but won't commit the so far collected usage data, and this in turn will commit all of the usage to the "last selected spool", which might not be what the user wanted.
To fix that, I'll try to replicate what SpoolManager plugin did in this situation - ask the user what should be done with the usage data until this point (whether to commit that to the previously selected spool, or commit to the new spool; or even cancel the spool selection change).

I think this answers your question, but if there's something I've missed, please let me know.

**Comment by evilDave** [comment]
I was going to begin another discussion but there is only this one there and it's basically the same topic.

I currently have different filaments set up in my slicer (orca) and filaments and spools in spoolman which are available to the octoprint plugin. So when I change colours I change it on the actual printer, then I change it in the slicer, slice and send, then I change it in octoprint via spoolman. (actually, most of the time I do a few prints sending data to the wrong spool for a while and then change over...)

I was wondering if you think it might be a good idea to do something like send a custom gcode from the slicer when the spool is selected. There is someone working on spoolman integration in orca, so they will be getting the filament ids and one will be able to map the slicer filament to the spool id. I think their integration then is to send the used length to spoolman upon uploading the job. What I think might be better is for the slicer to add custom gcode indicating which spool has been selected, which (I am assuming) can be picked up by the spoolman plugin to select the correct spool, and switch over to sending usage data to spoolman for that spool.

This way, the job of selecting a spool can be done once, and the usage data can be sent by someone who knows it has actually been used. I think this would continue to work for manual filament changes (though automatically) and could be made to work for automatic filament or tool changes becuase the filament/spool data will have been provided via the slicer.
